### ISSUE TITLE
Cannot create vmwarefusion machine if Fusion installed outside of /Applications

### ISSUE THREAD

**Issue by dleehr** [docker/machine#685]
I use [Homebrew Cask](http://caskroom.io) to install many binary applications like VMware Fusion. It places the .app bundles in `/opt/homebrew-cask` and symlinks them into `~/Applications`. As such, the `vmrun` command referenced by machine is not at the expected path.

The issue is that this command fails silently, and machine continues to wait for the VM to spin up, providing no useful feedback.

Ideally, machine could find `vmrun` on its own, but simply acknowledging/reporting the reason for the failure would be a good first step.

docker-machine version 0.1.0
Mac OS X 10.10.2

Debug output below

```
$ docker-machine -D create  -d vmwarefusion fusion
INFO[0000] Creating SSH key...                          
DEBU[0000] executing: /usr/bin/ssh-keygen ssh-keygen -t rsa -N  -f /Users/username/.docker/machine/machines/fusion/id_rsa
 
Generating public/private rsa key pair.
Your identification has been saved in /Users/username/.docker/machine/machines/fusion/id_rsa.
Your public key has been saved in /Users/username/.docker/machine/machines/fusion/id_rsa.pub.
The key fingerprint is:
xx:xx:xx:xx:xx:xx:xx:xx:xx:xx:xx:xx:xx:xx:xx:xx username@localhost
The key's randomart image is:
+--[ RSA 2048]----+
|                 |
|                 |
|                 |
|                 |
|                 |
|                 |
|                 |
|                 |
|                 |
+-----------------+
INFO[0000] Creating VM...                               
DEBU[0000] executing: /Applications/VMware Fusion.app/Contents/Library/vmrun start /Users/username/.docker/machine/machines/fusion/fusion.vmx nogui 
INFO[0000] Waiting for VM to come online...             
DEBU[0000] Not there yet 1/60, error: couldn't find MAC address in VMX file /Users/username/.docker/machine/machines/fusion/fusion.vmx 
DEBU[0002] Not there yet 2/60, error: couldn't find MAC address in VMX file /Users/username/.docker/machine/machines/fusion/fusion.vmx 
...
DEBU[0118] Not there yet 60/60, error: couldn't find MAC address in VMX file /Users/username/.docker/machine/machines/fusion/fusion.vmx 
ERRO[0120] Error creating machine: Machine didn't return an IP after 120 seconds, aborting 
WARN[0120] You will want to check the provider to make sure the machine and associated resources were properly removed. 
FATA[0120] Error creating machine                       
```

**Comment by dleehr** [comment]
On Mac OS X, mdfind can report the location of an app:

    $ mdfind 'kMDItemContentType == "com.apple.application-bundle" && kMDItemFSName = "VMware Fusion.app"'
    /opt/homebrew-cask/Caskroom/vmware-fusion/7.1.0-2314774/VMware Fusion.app

**Comment by sthulb** [comment]
+1 for detection.

**Comment by frapposelli** [comment]
Thanks for the tip, I'm actively working on a new version of the driver for
our hosted products (Fusion and Workstation), I will make sure to address
this.

**Comment by ehazlett** [comment]
thx @frapposelli

**Comment by nathanleclaire** [comment]
@frapposelli If the binary name is always the same, Golang's `exec.LookPath` is your friend.

**Comment by frapposelli** [comment]
@nathanleclaire isn't `exec.LookPath` looking only in `$PATH`? `vmrun` is usually buried into VMware specific directories.

**Comment by nathanleclaire** [comment]
@frapposelli Ah, yes, if it's not in `PATH` or equivalent you're going to have to look for it on your own.

**Comment by berinle** [comment]
@frapposelli  Any word on this? Using the latest version of docker machine (docker-machine version 0.2.0 (8b9eaf2)) and still running into the same issue. Thanks

**Comment by sthulb** [comment]
FWIW, I always install via cask and it works fine. Although, the vmrun binary is symlinked to `/usr/local/bin`

**Comment by berinle** [comment]
@sthulb thanks, but that is not the case for me. VM Fusion is installed with the standard installer from vmware and its located in the /Applications directory

**Comment by sthulb** [comment]
@berinle I guess the short term solution is to symlink the vmrun into your path - or add `/Applications/VMware Fusion.app/Contents/Library` to your `PATH` env var

**Comment by frapposelli** [comment]
If Fusion is installed in the default path it should work out of the box.

**Comment by berinle** [comment]
Hmmm, not joy on this end.

docker-machine create -d vmwarefusion --vmwarefusion-boot2docker-url https://github.com/cloudnativeapps/boot2docker/releases/download/v1.6.0-vmw/boot2docker-1.6.0-OAvmw.iso test
INFO[0000] Downloading boot2docker.iso from https://github.com/cloudnativeapps/boot2docker/releases/download/v1.6.0-vmw/boot2docker-1.6.0-OAvmw.iso...
INFO[0000] Creating SSH key...                        
INFO[0000] Creating VM...                             
INFO[0000] Starting test...                           
INFO[0000] Waiting for VM to come online...           
ERRO[0120] Error creating machine: Machine didn't return an IP after 120 seconds, aborting
WARN[0120] You will want to check the provider to make sure the machine and associated resources were properly removed.
FATA[0120] Error creating machine

docker-machine ls
ERRO[0000] error getting URL for host wpm: IP not found for MAC 00:0c:29:b4:fd:78 in DHCP leases
ERRO[0000] error getting URL for host test: IP not found for MAC 00:0c:29:d9:f5:ab in DHCP leases
NAME   ACTIVE   DRIVER         STATE     URL   SWARM
test            vmwarefusion   Running
wpm             vmwarefusion   Running
➜

**Comment by frapposelli** [comment]
@berinle that is a different problem, please create a new issue and post the machine creation log adding `-D` to the command.

**Comment by berinle** [comment]
@frapposelli ok, filed a new issue. #1177. Thanks. I'll do some more digging as well. It's puzzling to say the least

**Comment by abacon** [comment]
Hm — I'm getting the same thing even with vmrun symlinked into /usr/local/bin (& installed via cask). Any thoughts?

**Comment by dleehr** [comment]
@abacon The path to `vmrun` is [hard coded](https://github.com/docker/machine/blob/master/drivers/vmwarefusion/vmrun_darwin.go#L19) to `/Applications/VMware Fusion.app/Contents/Library/vmrun`

**Comment by ehazlett** [comment]
@frapposelli perhaps we could use `os.LookPath` for `vmrun`?

**Comment by sthulb** [comment]
(Copy from another #1322) This has some merit. Not everyone puts VMware fusion in the default place or uses home brew to install. Although, for the users that don't install it in either place, I'm unsure if we should bother, instead we should advise them to add it to their path manually.

**Comment by fgimian** [comment]
Same problem here folks, any possibility of a fix on the horizon? :smile: 

Thanks heaps!
Fotis

**Comment by fgimian** [comment]
Little trick I used to get me across the line for now :smile: 

```bash
ln -s /opt/homebrew-cask/Caskroom/vmware-fusion/7.1.2-2779224/VMware\ Fusion.app/ /Applications
```

Cheers
Fotis

**Comment by nathany** [comment]
Thanks @fgimian. However I'm still not able to start Vmware Fusion:

```console
❯ brew install docker-machine
==> Downloading https://homebrew.bintray.com/bottles/docker-machine-0.3.0.yosemi
######################################################################## 100.0%
==> Pouring docker-machine-0.3.0.yosemite.bottle.1.tar.gz
🍺  /usr/local/Cellar/docker-machine/0.3.0: 5 files, 16M
```

```console
❯ brew cask install vmware-fusion
==> Downloading https://download3.vmware.com/software/fusion/file/VMware-Fusion-
######################################################################## 100.0%
==> Symlinking App 'VMware Fusion.app' to '/Users/nathany/Applications/VMware Fu
==> Symlinking Binary 'ovftool' to '/usr/local/bin/ovftool'
==> Symlinking Binary 'vmnet-cfgcli' to '/usr/local/bin/vmnet-cfgcli'
==> Symlinking Binary 'vmnet-cli' to '/usr/local/bin/vmnet-cli'
==> Symlinking Binary 'vmrun' to '/usr/local/bin/vmrun'
==> Symlinking Binary 'vmware-vdiskmanager' to '/usr/local/bin/vmware-vdiskmanag
🍺  vmware-fusion staged at '/opt/homebrew-cask/Caskroom/vmware-fusion/7.1.2-2779224' (3285 files, 708M)
```

```
ln -s /opt/homebrew-cask/Caskroom/vmware-fusion/7.1.2-2779224/VMware\ Fusion.app/ /Applications
```

```
❯ docker-machine --debug start dev 
command=start machine=dev
executing: /Applications/VMware Fusion.app/Contents/Library/vmrun list
Starting dev...
executing: /Applications/VMware Fusion.app/Contents/Library/vmrun start /Users/nathany/.docker/machine/machines/dev/dev.vmx nogui
Mounting Shared Folders...
executing: /Applications/VMware Fusion.app/Contents/Library/vmrun -gu docker -gp tcuser runScriptInGuest /Users/nathany/.docker/machine/machines/dev/dev.vmx /bin/sh sudo mkdir /Users && sudo mount -t vmhgfs .host:/Users /Users
executing: /Applications/VMware Fusion.app/Contents/Library/vmrun list
executing: /Applications/VMware Fusion.app/Contents/Library/vmrun list
...
Maximum number of retries (60) exceeded
```

```console
"/Applications/VMware Fusion.app/Contents/Library/vmrun" list
Total running VMs: 0
```

**Comment by fgimian** [comment]
Cool mate.  I was going to suggest making sure you open VMware Fusion and entering your license before creating the machine, otherwise it may not work.

**Comment by fgimian** [comment]
Hey guys, I have attempted to fix this in my branch at https://github.com/fgimian/machine/tree/vmware-cask-support

I'm just testing it out to see if it works as expected but I welcome others to try it out too so I can submit it :smile: 

This is my first dabble in Go, so please let me know if I did anything wrong

Cheers
Fotis

**Comment by fgimian** [comment]
Had some dramas building a new docker machine to fully test this pull request but I have tested the altered code in a standalone script and it works as expected.

I'd welcome folks who can assist to give it a shot and let me know how it looks.

Cheers
Fotis

**Comment by jclagache** [comment]
I comment this issue because i am in this case too.
The easiest workaround is to symlink the harcoded paths, for example : 
```
ln -s /opt/homebrew-cask/Caskroom/vmware-fusion/7.1.0-2314774/VMware\ Fusion.app/ /Applications
 ```
I dug deeper to correct this behavior and i propose to do similarly as the virtualbox driver with a kind of "command find function": 
````
var (
	vmrunbin    = setVmwareCmd("vmrun")
	vdiskmanbin = setVmwareCmd("vmware-vdiskmanager")
)
// detect the vmrun and vmware-vdiskmanager cmds' path if needed
func setVmwareCmd(cmd string) string {
	if path, err := exec.LookPath(cmd); err == nil {
		return path
	}
	usr, _ := user.Current()
    usr_vmfusionapp := filepath.Join(usr.HomeDir, "Applications/VMware Fusion.app")
    if _, err := os.Stat(usr_vmfusionapp); !os.IsNotExist(err) {
    	return filepath.Join(usr_vmfusionapp, "Contents/Library/", cmd)
    }
    return filepath.Join("/Applications/VMware Fusion.app/Contents/Library/", cmd) 
}
```
let me know if it suits before i make the PR.

**Comment by nathanleclaire** [comment]
@frapposelli On this issue, why not use `exec.LookPath` and then check the hardcoded directory if the command is not in `PATH`?

**Comment by jclagache** [comment]
@nathanleclaire I agree : examine the system path and the harcoded directory is enough. Brew cask links and [will continue to link](https://github.com/caskroom/homebrew-cask/issues/13201) executables in the system path by default.

**Comment by dleehr** [comment]
This has been fixed by #1812, thanks!
